filetree/tree.go

```diff
@@ -35,8 +36,95 @@ func NewFileTree() (tree *FileTree) {
 	return tree
 }
 
+type renderParams struct{
+	node *FileNode
+	spaces []bool
+	childSpaces []bool
+	showCollapsed bool
+	isLast bool
+}
+
+func (tree *FileTree) renderStringTreeBetween(startRow, stopRow int, showAttributes bool) string {
+	// generate a list of nodes to render
+	var params []renderParams = make([]renderParams,0)
+	var result string
+
+	// visit from the front of the list
+	var paramsToVisit = []renderParams{ renderParams{node: tree.Root, spaces: []bool{}, showCollapsed: false, isLast: false} }
+	for currentRow := 0; len(paramsToVisit) > 0 && currentRow <= stopRow; currentRow++ {
+		// pop the first node
+		var currentParams renderParams
+		currentParams, paramsToVisit = paramsToVisit[0], paramsToVisit[1:]
+
+		// take note of the next nodes to visit later
+		var keys []string
+		for key := range currentParams.node.Children {
+			keys = append(keys, key)
+		}
+		sort.Strings(keys)
+
+		var childParams = make([]renderParams,0)
+		for idx, name := range keys {
+			child := currentParams.node.Children[name]
+			// don't visit this node...
+			if child.Data.ViewInfo.Hidden || currentParams.node.Data.ViewInfo.Collapsed {
+				continue
+			}
+
+			// visit this node...
+			isLast := idx == (len(currentParams.node.Children) - 1)
+			showCollapsed := child.Data.ViewInfo.Collapsed && len(child.Children) > 0
+
+			// completely copy the reference slice
+			childSpaces := make([]bool, len(currentParams.childSpaces))
+			copy(childSpaces, currentParams.childSpaces)
+
+			if len(child.Children) > 0 && !child.Data.ViewInfo.Collapsed {
+				childSpaces = append(childSpaces, isLast)
+			}
+
+			childParams = append(childParams, renderParams{
+				node: child,
+				spaces: currentParams.childSpaces,
+				childSpaces: childSpaces,
+				showCollapsed: showCollapsed,
+				isLast: isLast,
+			})
+		}
+		// keep the child nodes to visit later
+		paramsToVisit = append(childParams, paramsToVisit...)
+
+		// never process the root node
```

Comment:
I don't understand why you never process the root. Does this mean that the line `/` in the file system never shoes up in the output?

Reply:
Indeed, this is so the root dir never shows up in the filetree and we also don't have any information on it. The tar you get from the docker save command has the contents of what in the images root dir, but no info on `/` itself. I couldn't see a reason to include it in the tree other than to hide all files... so I left it out.

However, from a data structure point of view keeping a single root node simplifies how you traverse the tree (instead of having multiple root nodes you have one, and the children are the "real" start to the tree).

Downside is preventing weird one-off bugs :/